cmd/clusterctl/pkg/client/cluster/inventory.go

```diff
@@ -294,3 +308,115 @@ func (p *inventoryClient) GetDefaultProviderNamespace(provider string) (string,
 	// There is no provider or more than one namespace for this provider; in both cases, a default provider namespace cannot be decided.
 	return "", nil
 }
+
+func (p *inventoryClient) GetManagementGroups() ([]ManagementGroup, error) {
+	providers, err := p.List()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if there are core providers with overlapping watching namespaces, if yes, it is not possible to define the management groups, so return error
+	if err := checkOverlappingCoreProviders(providers); err != nil {
+		return nil, err
+	}
+
+	// check if a Bootstrap/ControlPlane/Infrastructure providers can't be combined with any core provider or
+	// if any Bootstrap/ControlPlane/Infrastructure provider is watching for objects controlled by more than one core provider.
+	// In both cases it is not possible to define the management groups, so return error
+	if err := checkOverlappingProviders(providers); err != nil {
+		return nil, err
+	}
+
+	managementGroups := []ManagementGroup{}
+	for _, coreProvider := range providers {
+		if coreProvider.Type != string(clusterctlv1.CoreProviderType) {
+			continue
+		}
+
+		group := ManagementGroup{CoreProvider: coreProvider}
+		for _, provider := range providers {
+			if coreProvider.WatchedNamespace == "" || coreProvider.WatchedNamespace == provider.WatchedNamespace || provider.WatchedNamespace == "" {
+				group.Providers = append(group.Providers, provider)
+			}
+		}
+		managementGroups = append(managementGroups, group)
+	}
+
+	return managementGroups, nil
+}
+
+// checkOverlappingCoreProviders checks if there are core providers with overlapping watching namespaces, if yes, return error e.g.
+// cluster-api in capi-system watching all namespaces and another cluster-api in capi-system2 watching capi-system2 (both are watching capi-system2)
+// NB. This should not happen because init prevent the users to do so, but nevertheless we are double checking this before upgrades.
+func checkOverlappingCoreProviders(providers []clusterctlv1.Provider) error {
+	for _, provider := range providers {
+		// if the provider isn't a core provider, skip it
+		if provider.Type != string(clusterctlv1.CoreProviderType) {
+			continue
+		}
+
+		// check for overlapping core providers
+		for _, providerToCompare := range providers {
+			// if the provider isn't a core provider, skip it
+			if providerToCompare.Type != string(clusterctlv1.CoreProviderType) {
+				continue
+			}
+
+			// if the provider to compare is the same of the outer provider, skip it
+			if provider.Namespace == providerToCompare.Namespace && provider.Name == providerToCompare.Name {
```

Comment:
These checks should probably go in methods on the `Provider` struct, here it could be `Equals(other Provider)`